### PR TITLE
Pass .chunk/rechunk calls through for chunked arrays without ChunkManagers

### LINKED ISSUE #8733: A basic default ChunkManager for arrays that report their own chunks
### Is your feature request related to a problem?

I'm creating duckarrays for various file backed datastructures for mine that are naturally "chunked". i.e. different parts of the array may appear in completely different files.

Using these "chunks" and the "strides" algorithms can better decide on how to iterate in a convenient manner.

For example, an MP4 file's chunks may be defined as being delimited by I frames, while images stored in a TIFF may be delimited by a page.

So for me, chunks are not so useful for parallel computing, but more for computing locally and choosing the appropriate way to iterate through a large arrays (TB of uncompressed data).

### Describe the solution you'd like

I think a default Chunk manager could simply implement `compute` as `np.asarray` as a default instance, and be a catchall to all other instances.

Advanced users could then go in an reimplement their own chunkmanager, but I was unable to use my duckarrays that incldued a `chunk` property because they weren't associated with any chunk manager.

Something as simple as:

```patch
diff --git a/xarray/core/parallelcompat.py b/xarray/core/parallelcompat.py
index c009ef48..bf500abb 100644
--- a/xarray/core/parallelcompat.py
+++ b/xarray/core/parallelcompat.py
@@ -681,3 +681,26 @@ class ChunkManagerEntrypoint(ABC, Generic[T_ChunkedArray]):
         cubed.store
         """
         raise NotImplementedError()
+
+
+class DefaultChunkManager(ChunkMangerEntrypoint):
+    def __init__(self) -> None:
+        self.array_cls = None
+
+    def is_chunked_array(self, data: Any) -> bool:
+        return is_duck_array(data) and hasattr(data, "chunks")
+
+    def chunks(self, data: T_ChunkedArray) -> T_NormalizedChunks:
+        return data.chunks
+
+    def compute(self, *data: T_ChunkedArray | Any, **kwargs) -> tuple[np.ndarray, ...]:
+        raise tuple(np.asarray(d) for d in data)
+
+    def normalize_chunks(self, *args, **kwargs):
+        raise NotImplementedError()
+
+    def from_array(self, *args, **kwargs):
+        raise NotImplementedError()
+
+    def apply_gufunc(self, *args, **kwargs):
+        raise NotImplementedError()

```

### Describe alternatives you've considered

I created my own chunk manager, with my own chunk manager entry point.

Kinda tedious...

### Additional context

It seems that this is related to: https://github.com/pydata/xarray/pull/7019

### PR DESCRIPTION
**Pull request by TomNicholas**
Basically implements @dcherian 's suggestion from https://github.com/pydata/xarray/issues/8733#issuecomment-2251861380:

> IIUC this mostly gets resolved if the ChunkManager is less greedy and doesn't trigger on the existence of .chunks but is instead triggered on matching an allowlist of registered chunk array types.

Needed to fix https://github.com/zarr-developers/VirtualiZarr/pull/199#issuecomment-2244586390.

The actual fix is in just the first two commits, the rest is defining a new `has_chunkmanager` function and using that everywhere to distinguish between arrays that have `.chunks` (e.g. `virtualizarr.ManifestArray`) and arrays that actually need to call out to a ChunkManager (i.e. dask/cubed).

- [x] Closes #8733, superceding #8739
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] ~~New functions/methods are listed in `api.rst`~~

### CODE REVIEW

File: xarray/namedarray/core.py [hunk 1]
```diff
@@ -819,11 +819,18 @@ def chunk(
...
-
         data_old = self._data
-        if chunkmanager.is_chunked_array(data_old):
-            data_chunked = chunkmanager.rechunk(data_old, chunks)  # type: ignore[arg-type]
+        if is_chunked_array(data_old):
+            print(f"problematic chunks = {chunks}")
+            # if is_dict_like(chunks) and chunks != {}:
+            #     chunks = tuple(chunks.get(n, s) for n, s in enumerate(data_old.shape))  # type: ignore[assignment]
+
+            print(f"hopefully normalized chunks = {chunks}")
```
Comment (TomNicholas) [inline xarray/namedarray/core.py:828]: This is really irritating - if I keep these lines commented out then my `test_rechunk` on the `DummyChunkedArray` fails. But if I uncomment these lines (therefore doing exactly what happens in the other branch of the `if is_chunked_array(data_old):` statement) then dask rechunk tests fail! 

There are so many possible valid argument types for `chunks` here, some of which are `dicts` but completely different, e.g. `{0: (2, 3)}` vs `{'x': (2, 3)}`.

It would be much nicer for all possible `chunks` to go through a single `normalize_chunks` function, but I'm getting confused even trying to work out what the current behaviour is.

The `ChunkManager` has a `.normalize_chunks` method, to call out to `dask.array.normalize_chunks`. Cubed [vendors this function](https://github.com/cubed-dev/cubed/blob/69e9f94241f3b64924f52ec50e0548c24f79a8b6/cubed/vendor/dask/array/core.py#L103) too, so perhaps instead xarray should vendor `dask.array.normalize_chunks` and remove it from the `ChunkManager` class?

File: xarray/tests/test_parallelcompat.py [hunk 2]
```diff
@@ -147,6 +157,35 @@ def register_dummy_chunkmanager(monkeypatch):
...
+        na: NamedArray = NamedArray(data=dummy_arr, dims=["x"])
+        rechunked_na = na.chunk(chunks={"x": (2,)})
+        assert isinstance(rechunked_na.data, DummyChunkedArray)
+        assert rechunked_na.data.chunks == ((2,),)
+        assert rechunked_na.chunksizes == {"x": (2,)}
+
+    def test_computation(self) -> None:
+        dummy_arr = DummyChunkedArray(shape=(4,), chunks=((1,),))
+        na: NamedArray = NamedArray(data=dummy_arr, dims=["x"])
+        na.mean()
```
Comment (TomNicholas) [inline xarray/tests/test_parallelcompat.py:186]: Not entirely sure what the intended behaviour should be here. This test tests what happens if you try to compute an array that has `.chunks` but is not registered via any chunkmanager. 

In virtualizarr's case this situation should just raise immediately because `ManifestArray`s are not computable, so from virtualizarr's PoV it doesn't really matter what happens here.

@hmaarrfk what is the preferred behaviour for your chunked arrays?

I guess if it does attempt to pass computation through here that could cause issues when computing on a cubed array with cubed-xarray not installed... (That scenario can't happen for dask because the equivalent DaskManager (i.e. dask-xarray) is effectively bundled inside xarray.)

File: xarray/tests/test_parallelcompat.py [hunk 3]
```diff
@@ -19,6 +20,7 @@
 from xarray.tests import has_dask, requires_dask
 
 
+# TODO can I subclass the chunkedduckarray protocol here?
```
Comment (TomNicholas) [inline xarray/tests/test_parallelcompat.py:23]: @Illviljan I hope I'm using all your cool duckarray type protocols correctly!

File: xarray/namedarray/pycompat.py [hunk 4]
```diff
@@ -89,7 +90,23 @@ def mod_version(mod: ModType) -> Version:
...
+
+    try:
+        get_chunked_array_type(x)
+    except TypeError as e:
+        if str(e).startswith("Could not find a Chunk Manager which recognises type"):
+            return False
+        elif str(e) == "Expected a chunked array but none were found":
+            return False
+        else:
+            raise  # something else went wrong
```
Comment (TomNicholas) [inline xarray/namedarray/pycompat.py:107]: This might be a code smell, in which case `has_chunkmanager`, `guess_chunkmanager`, and `get_chunked_array_type` should be refactored.

File: xarray/coding/strings.py [hunk 5]
```diff
@@ -183,7 +183,7 @@ def char_to_bytes(arr):
         # can't make an S0 dtype
         return np.zeros(arr.shape[:-1], dtype=np.bytes_)
 
-    if is_chunked_array(arr):
+    if is_chunked_array(arr) and has_chunkmanager(arr):
```
Comment (TomNicholas) [inline xarray/coding/strings.py:186]: This `is_chunked_array(arr) and has_chunkmanager(arr)` pattern becomes necessary because we are now considering the possibility that `is_chunked_array(arr) == True` but `has_chunkmanager(arr) == False`, whereas previously these were assumed to always be consistent.

File: xarray/tests/test_parallelcompat.py (hunk 2 above)
Comment (hmaarrfk) [reply xarray/tests/test_parallelcompat.py:186]: The users of my array (the rest of our team) feels like all these should do something.

It might make things REALLY slow, but I feel like `mean` should compute.... Your chunked array should know how best to compute it for itself. How should it compute intermediate results? In what order should it go through the array.

Comment (TomNicholas) [reply xarray/tests/test_parallelcompat.py:186]: > It might make things REALLY slow, but I feel like mean should compute....

Yes I agree.

> How should it compute intermediate results? In what order should it go through the array.

I'm not sure I understand you here.

> I guess if it does attempt to pass computation through here that could cause issues when computing on a cubed array with cubed-xarray not installed...

Thinking about this more I don't think it's a big deal. Not recognising a chunked array type will just mean that xarray falls back to calling numpy functions on it (e.g. `da.mean()` will call `np.mean(arr)`), which will call `__array__` on the underlying type, coercing it to numpy, and in the case of cubed arrays, simply eagerly computing it.

File: xarray/coding/strings.py (hunk 5 above)
Comment (headtr1ck) [reply xarray/coding/strings.py:186]: But doesn't  `has_chunkmanager(arr) == True` imply `is_chunked_array(arr) == True`?

Comment (TomNicholas) [reply xarray/coding/strings.py:186]: @headtr1ck I got a notification saying you commented saying

> But doesn't `has_chunkmanager(arr) == True` imply `is_chunked_array(arr) == True`?

(But I can't find your comment.)

It's a good question though. I _think_ there are some array types that don't define a `.chunks` where you might still want to use other `ChunkManager` methods.

In particular JAX is interesting - it has a top-level [`pmap`](https://jax.readthedocs.io/en/latest/_autosummary/jax.pmap.html) function which applies a function over multiple axes of an array similar to `apply_gufunc`. It distributes computation, but not over `.chunks` (which JAX doesn't define), instead over a global variable [`jax.local_device_count`](https://jax.readthedocs.io/en/latest/_autosummary/jax.local_device_count.html#jax.local_device_count). 

This is why I think we should rename `ChunkManager` to `ComputeManager`.

cc @alxmrs

Comment (headtr1ck) [reply xarray/coding/strings.py:186]: > @headtr1ck I got a notification saying you commented saying
> 
> > But doesn't `has_chunkmanager(arr) == True` imply `is_chunked_array(arr) == True`?
> 
> (But I can't find your comment.)
> 
> It's a good question though. I _think_ there are some array types that don't define a `.chunks` where you might still want to use other `ChunkManager` methods.
> 

I came to the same conclusion, that's why I deleted the comment, sry.

File: xarray/tests/test_parallelcompat.py (hunk 2 above)
Comment (hmaarrfk) [reply xarray/tests/test_parallelcompat.py:186]: > cubed arrays, simply eagerly computing it.

maybe cubed would help us in our lazy arrays.

Comment (hmaarrfk) [reply xarray/tests/test_parallelcompat.py:186]: Me personally that is

File: xarray/coding/strings.py (hunk 5 above)
Comment (TomNicholas) [reply xarray/coding/strings.py:186]: No worries! I prefer to leave all my half-baked thoughts in the open and double or triple-post 😅 If you were wondering it then other people will definitely have the same question!

> This is why I think we should rename `ChunkManager` to `ComputeManager`.

I could leave this to a second PR, to isolate the breaking changes.

File: xarray/tests/test_parallelcompat.py (hunk 2 above)
Comment (TomNicholas) [reply xarray/tests/test_parallelcompat.py]: > Thinking about this more I don't think it's a big deal.

You also kinda have to go out of your way to even created an xarray-wrapped cubed array without cubed-xarray installed, because you can only use `open_dataset` and `.chunk` to get cubed arrays if you have cubed-xarray installed.

> maybe cubed would help us in our lazy arrays.

Maybe! The [`cubed.Plan`](https://cubed-dev.github.io/cubed/computation.html#plan) model is super nice.

File: xarray/namedarray/core.py (hunk 1 above)
Comment (TomNicholas) [reply xarray/namedarray/core.py]: > `test_rechunk` on the `DummyChunkedArray` fails

This was actually mostly my fault for having a bug in that test, fixed by https://github.com/pydata/xarray/pull/9286/commits/0296f92d207afb7319b4e31ff03a76c145ddb251.

> It would be much nicer for all possible chunks to go through a single normalize_chunks function

But there is still some unnecessary complexity that would be nice to remove. The main reason why the weird `is_dict_like(chunks):` sections that turn dicts of chunks into tuples are currently needed is because of this bug in `dask.array.core.normalize_chunks` https://github.com/dask/dask/issues/11261. Otherwise we could just use that.

(If we do just use that we should perhaps vendor it though - as cubed does already).

Comment (TomNicholas) [reply xarray/namedarray/core.py]: I managed to sort this all out, so now everything goes through `dask.array.core.normalize_chunks`, which is much neater.

Question is now do I:
1) Vendor `dask.array.core.normalize_chunks` (like cubed does), and use the vendored version no matter which `ChunkManager` is called
2) Make all chunkmanagers define a normalize_chunks method and refer to that (what the `main` code currently does).

I think we actually have to do (1), because we now have a codepath which will try to call `normalize_chunks` even on chunked arrays that do not define a chunkmanager. But we want to vendor it without introducing any more dependencies (e.g. `toolz`).

Comment (TomNicholas) [reply xarray/namedarray/core.py]: @dcherian I would appreciate your input on this vendoring question before I move ahead with it ^

Comment (dcherian) [reply xarray/namedarray/core.py]: Vendor it! Sorry for the delay. We can generalize if it's ever needed

File: xarray/coding/strings.py (hunk 5 above)
Comment (mjwillson) [reply xarray/coding/strings.py:186]: @TomNicholas FYI JAX does now support something a bit like chunking via sharding of `jax.Array`, there's a good summary here: https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html
IIUC this is now preferred over `pmap`.
